core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/BadEventTransformer.scala

```diff
@@ -62,5 +61,66 @@ class BadEventTransformer
    * @return Line as an EmitterJsonInput
    */
   def consumeLine(line: String): EmitterJsonInput =
-    fromClass(line -> JsonRecord(JObject(JField("source", JString(line))), None).success)
+    fromClass(line -> handleIgluJson(line).map(json => JsonRecord(json, None)))
+
+}
+
+object BadEventTransformer {
+
+  /** All transformations to avoid union types within `iglu:com.snowplowanalytics.snowplow.badrows` */
+  val fixes: List[Json => Json] = List(
+    root.obj.modify(renameField("failure")),
+    root.obj.modify(renameField("payload")),
+    root.payload.raw.obj.modify(serializeField("parameters")),
+    root.failure.obj.modify(renameField("error")),
+    root.failure.obj.modify(renameField("message")),
+    root.failure.messages.each.obj.modify(renameField("error")),
+    root.failure.messages.each.obj.modify(serializeField("expectedMapping")),
+    root.failure.messages.each.obj.modify(serializeField("json")),
+    root.failure.messages.each.message.obj.modify(renameField("error")),
+    root.failure_list.each.obj.modify(renameField("error")),
+    root.failure_list.each.obj.modify(serializeField("value"))
+  )
+
+  /** Attempt to handle self-describing JSON and fix bad rows union types */
+  def handleIgluJson(row: String): ValidatedNel[String, Json] =
+    parse(row) match {
+      case Right(json) =>
+        json.as[SelfDescribingData[Json]] match {
+          case Right(SelfDescribingData(schema, data)) if schema.vendor.contains("badrows") =>
+            SelfDescribingData(schema, transform(data)).asJson.validNel
+          case Right(_) =>
+            json.validNel
+          case Left(_) => json.validNel
+        }
+      case Left(_) => s"BadEventTransformer cannot parse $row as JSON".invalidNel
+    }
+
+  /**
+   * Apply every known `BadRow` fix to a resulting JSON
+   * @param badRow BadRow payload *without* self-describing meta
+   * @return fixed JSON without union types
+   */
+  def transform(badRow: Json): Json =
+    fixes.foldLeft(badRow) { (data, fix) =>
+      fix.apply(data)
+    }
+
+  /** Append a type suffix to a key */
+  def renameField(field: String)(jsonObject: JsonObject): JsonObject =
+    jsonObject(field) match {
+      case Some(value) if value.isString =>
+        jsonObject.remove(field).add(field ++ "_str", value)
+      case Some(value) if value.isArray =>
+        jsonObject.remove(field).add(field ++ "_list", value)
+      case _ => jsonObject
+    }
+
+  /** Stringify JSON value to avoid potential index explosion */
+  def serializeField(field: String)(jsonObject: JsonObject): JsonObject =
+    jsonObject(field) match {
+      case Some(value) if !value.isString =>
+        jsonObject.add(field, Json.fromString(value.noSpaces))
+      case _ => jsonObject
+    }
```

Comment:
Brilliant!

Reply:
The implementation was ~5 times longer before I used circe optics. FP FTW!